Desktop calendar

Wilker0725 opened the pull request:
**Changes**
Description of changes...

- delete branch

<!-- Remove if not required -->
**Changelog entry**

- [BREAKING | WARNING] Description of change in `desktop-calendar`

**Additional information**

- You may refer to this [ticket](url)
- Any other information you would like to include

qroll commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,394 @@
+    const generateWeek = useCallback((day: Dayjs): Date[] => {
+        const dates: Date[] = [];
+        for (let i = 0; i < 7; i++) {
+            const date = day.clone().add(i, "day").toDate();
```
so far the component has been manipulating the dates with `Dayjs`, so why convert to `Date` here?

Wilker0725 replied:
`Dayjs` was a object from the dayJs package. In line 258, `toDate()` will be formatted to `Fri Feb 03 2023 13:53:27 GMT+0800 (Malaysia Time)` Date String instead of DayJs object

Wilker0725 replied:
Handle a Raw Date format is easy to readable and do other function from the raw Date format. 

The DayJs is an object from the package. Need to be transform back to raw format in other function.

qroll commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,394 @@
+            .replaceAll("/", "-");
+
+        // YYYY-MM-DD
+        const dateStartWithYear = revertedDateFormat(_dateStartWithDay);
```
refer to prev comment about converting `Dayjs` to `Date`. can use `dayjs.format()` directly

Wilker0725 replied:
Yes. Both way are workable.

qroll replied:
if there's no other use for `Date` other than formatting, I think we should stick to `Dayjs` within the component. then no need to convert back and forth

Wilker0725 replied:
My bad. check it again for the reason of using toDate().

Native Date string:  Thu Mar 02 2023 00:00:00 GMT+0800 (Malaysia Time). ----->  toDate()
Date string: dayjs().format("YYYY-MM-DD") or dayjs().format(). That's a string, not Date String. 

https://github.com/LifeSG/react-design-system/blob/3080b00782e8e1930a4212a2791803ed3223c723/src/date-picker/date-picker.tsx#L351

As this example the day will to handle in other way instead using the Native Browser Fn.
```
 calendarDate?.clone().toDate().getMonth() !== day.getMonth()
```

qroll reviewed (changes requested):
have some initial comments after looking through the parent calendar component:
- the mixed use of `Dayjs` and `Date` might be unnecessary
- the day, month and year generators can be pulled out into a util file with unit tests

Wilker0725 commented on src/index.ts:
```diff
@@ -6,6 +6,7 @@ export * from "./box-container";
 export * from "./breadcrumb";
 export * from "./button";
 export * from "./card";
+export * from "./calendar";
```
noted. let me check those again.

Wilker0725 commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,394 @@
+                </DropdownYear>
+            </HeaderDropdown>
+            <Views show={showView}>
+                <DayView>
```
As default <Views> is a parent element with 03 children which is <DayView>, <MonthView> and <YearView>.

<Views> will be control the view from the "showView" Value and styling it for the children view.

`
export const Views = styled.div<CalendarViewProps>`
    margin-left: auto;
    margin-right: auto;
    display: flex;
    flex-direction: column;

    ${(props) => {
        const _VIEWS = ["Day", "Month", "Year"] as View[];
        let styles = "";
        for (let i = 0; i < props.children.length; i++) {
            styles += `
                    > div:nth-child(${i + 1}) {
                        display: ${props.show === _VIEWS[i] ? "flex" : "none"} ;
                    }
                `;
        }
        return css`
            ${styles}
        `;
    }}
`

Wilker0725 commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,394 @@
+        setSelectedStartDate(value);
+    };
+
+    // =============================================================================
+    // Calendar Generate
+    // =============================================================================
```
The memoise hook use for prevent it do render in unnecessary update which may be the state been updated in other component which cause this calendar re-calculated again. correct me if i'm wrong to use the useMemo.

Wilker0725 replied:
Question:

The calendar will be generate depend on the `calendarDate`. You need a hook for the useCalendar.tsx file or ts file?

```
const firstDayOfTheMonth = useMemo(
    () => calendarDate.clone().startOf("month"),
    [calendarDate]
);

```

qroll replied:
it's a good idea to memoise the final calculated value. but since `firstDayOfTheMonth` and `firstDayOfFirstWeekOfMonth` are both derived from `calendarDate`, they can all be computed together

I'm proposing

```
// src/utils/calendar-helper.ts

const generateWeeksOfTheMonth = (calendarDay: Dayjs) => {
  const firstDayOfTheMonth = ...
  const firstDayOfFirstWeekOfMonth = ...
  const firstDayOfEachWeek = ...
  return firstDayOfEachWeek.map(...)
}

// src/calendar/calendar.tsx

const weeksOfTheMonth = useMemo(() => generateWeeksOfTheMonth(calendarDay), [calendarDay]);
```

can correct me if I missed out any dependencies

Wilker0725 replied:
Wa. you response the final answer. Thanks to guide me. I learn bit from you. 

I was thinking that was a useHook instead of ts file.

Wilker0725 commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,35 @@
+    | "selectedYear";
+
+export interface CalendarProps {
+    disabledDate?: string[];
```
wondering that why we needed undefined? to prevent someone use disabledDates with not value?

qroll replied:
it's a DS convention for to set `| undefined` for optional values for better type handling, also provides support for [exactOptionalPropertyTypes](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes)

Wilker0725 replied:
Thanks sharing that

Wilker0725 commented:
Thanks for your time again!

Updated:
1. Sort Lines Ascending 
2. Remove duplicated in /src/calendar/types
3. Update plural for disabledDate
4. Added `| undefined` for value in types
5. Import new Icon package from DS
6. Generate Calendar move into utils folder
7. Unit test for utils calendar-helper
8. Update Native Date to Dayjs format

qroll commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,33 @@
+export type VariantYear =
+    | "default"
+    | "currentYear"
+    | "nextDecaded"
```
I think you mean

```suggestion
    | "nextDecade"
```
can update all other instances of `decaded` too

qroll commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,348 @@
+                                        key={`day-${dayIndex}`}
+                                        className={classNames}
+                                    >
+                                        <LeftCell data-cell="left">
```
are these `data-cell` and `data-circle` attributes used for anything?

Wilker0725 replied:
`data-cell` and `data-circle` use in css styling

Wilker0725 replied:
Updated: 
1. Namespace and Naming for calendar-helper
2. Move `generateFirstDayOfEachWeek` and `generateWeek` outside of the main fn in calendar-helper file.
3. Remove `clone()` for dayjs
4. Create new component for `CalendarDay`
5. All `px` changed to `rem`
6. Use `Dayjs` in month and year component instead of `Date`
7. Remove `null` in initial mount in month calendar
8. Re-name to `nextDecade ` from `nextDecaded`
9. Move `generateMonth` and `generateDecadeOfYear` into calendar-helper
10. Added unit test for `generateMonth` and `generateDecadeOfYear` in calendar-helper

Pending:
1. Simplify for dayjs(calendarDate).month(i)
2. Simplify for dayjs(calendarDate).year(i)

Wilker0725 commented on src/calendar/calendar-month.tsx:
```diff
@@ -0,0 +1,109 @@
+    showView,
+    selectedStartDate,
+}: CalendarMonthProps) => {
+    const [monthDate, setMonthDate] = useState<string[]>([""]);
```
thanks pointing out for this one. 

After checked. i will removed line 74. it wouldn't run calculation for month calendar in initial mounted also.

My initial thoughts was not render month calendar in initial mounted for month calendar to improve performance.

Wilker0725 commented on src/calendar/calendar-month.tsx:
```diff
@@ -0,0 +1,109 @@
+        };
+    };
+
+    const generateMonth = () => {
```
Thank you for this one. 

but it return dayjs object from this  `dayjs("2023-10-31").month(2)` instead of the number of the month.

Other,  return **31** day from `dayjs("2023-10-31").month(2).daysInMonth()`;

qroll replied:
sorry my example should have been `dayjs("2023-10-31").month(1)` (starts counting from 0), then `daysInMonth()` will return 28 for February

anyway I just wanted to show how to generate 1 dayjs object for every month, then they can be transformed to whatever format you need afterwards

Wilker0725 replied:
you so awesome. Thanks sharing that. 

I do a stupid way. hahah

Wilker0725 replied:
Please check this. I guess i did a stupid way again.  

`const [yyyy, , dd] = calendarDate.format("YYYY-MM-DD").split("-");`, to know what's the today day.

```
  const generateMonth = () => {
        const months: Dayjs[] = [];
        const [yyyy, , dd] = calendarDate.format("YYYY-MM-DD").split("-");

        for (let i = 0; i < 12; i++) {
            let month: Dayjs;
            month = dayjs(calendarDate).month(i);
            const dayInMonth: number = month.daysInMonth();

            if (+dd > dayInMonth) {
                // get last day of the month
                month = dayjs(`${yyyy}-${Months[i]}-${dayInMonth}`);
            }

            months.push(dayjs(month));
        }

        setMonthDate(months);
    };
```

Wilker0725 replied:
Let me know what needed. It look like didn't simplified it. 

I needed `const [yyyy, , dd] = calendarDate.format("YYYY-MM-DD").split("-")` because of one scenario as

User select Feb from 2023-01-31, that line of code is track the selected day or today day.



```
export const generateMonth = (calendarDate: Dayjs) => {
        const months: Dayjs[] = [];

        const [yyyy, , dd] = calendarDate.format("YYYY-MM-DD").split("-");

        for (let i = 0; i < 12; i++) {
            let month: Dayjs;
            month = dayjs(calendarDate).month(i);
            const dayInMonth: number = month.daysInMonth();

            if (+dd > dayInMonth) {
                // get last day of the month
                const mmmm = month.format("MMMM");
                month = dayjs(`${yyyy}-${mmmm}-${dayInMonth}`);
            }

            months.push(dayjs(month));
        }

        return months;
    };
```

qroll replied:
this is what I had in mind (pseudo code), no need to handle months with less days since dayjs adjusts it automatically
```
for (i = 0 to 11) {
  monthForSelectedDay = calendarDate.month(i)
  months.push(monthForSelectedDay)
}
```

btw should be `generateMonths` since it returns an array

Wilker0725 replied:
Thanks again for your time to review it again. 

The logic been simplified a lot. thanks!

Wilker0725 commented:
Updated:

Namespace and Naming for calendar-helper
Move generateFirstDayOfEachWeek and generateWeek outside of the main fn in calendar-helper file.
Remove clone() for dayjs
Create new component for CalendarDay
All px changed to rem
Use Dayjs in month and year component instead of Date
Return null in initial mount in month calendar for day component mount.
Re-name to nextDecade  from nextDecaded
Move generateMonth and generateDecadeOfYear into calendar-helper
Added unit test for generateMonth and generateDecadeOfYear in calendar-helper
Pending:

Simplify for dayjs(calendarDate).month(i)
Simplify for dayjs(calendarDate).year(i)

Wilker0725 commented on src/calendar/calendar-year.tsx:
```diff
@@ -0,0 +1,96 @@
+        };
+    };
+
+    const generateDecadeOfYear = () => {
```
The `mm` and `dd` are required for the value user selected. I have not idea how to implement your suggestion into my code. 

Could you share more details?

https://github.com/LifeSG/react-design-system/blob/7f97c403c119ca95662deb94599bff5aa911607f/src/util/calendar-helper.ts#L54

qroll replied:
yup something like that. use dayjs to either set or add/subtract the years

if you prefer to use add/subtract, i would suggest something like this (pseudo-code)
```
year = calendarDate.year()
decade = Math.floor(yyyy / 10) * 10

base = calendarDate.year(decade)
prev = base.subtract(1, "year")
years = [prev, base]
for (i = 1 to 10) {
  years.push(base.add(i, "year"))
}
```

would prefer not doing `if (i === 0)` etc in loops because that makes it harder to understand the behaviour of each iteration

Wilker0725 replied:
Awesome. Learn a lot from you. Thanks!

Wilker0725 replied:
Great. You logic are awesome. Simple and understandable

Wilker0725 commented:
Updated:

1. Simplified month calendar with `calendar.month(i)`.
2. Simplified year calendar with `calendar.year(i)`.

qroll commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,132 @@
+                        transform: rotate(180deg);
+                    }
+                `;
+            case "Year":
```
identical to the previous case, so can merge

Wilker0725 replied:
previous case??

Wilker0725 replied:
ohh.. the prefix $?

qroll replied:
I meant in the switch case, Month and Year are handled the same way

so actually you can do
```
switch (...) {
  case "Month":
  case "Year":
    return css``
```

Wilker0725 replied:
Opps. Yes. they are same styling lol.

qroll commented on src/calendar/calendar-day.style.tsx:
```diff
@@ -0,0 +1,180 @@
+`;
+
+export const WeekCell = styled.div`
+    ${TextStyleHelper.getFontFamily("H6", "semibold")}
```
use `TextStyleHelper.getTextStyle` instead if you need the matching font-size

Wilker0725 replied:
Thanks for sharing this method. This is what i need.

qroll reviewed (changes requested):
thanks for updating! the main structure looks ok, the next comments I have are mostly for style and code cleanliness

key changes required:
- [ ] prefix styled component props with $
- [ ] organise imports in each file, and make the imports relative (e.g. `src/color` -> `../color`)
- [ ] check through for unnecessary use of `dayjs(...)`
- [ ] use `TextStyleHelper. getTextStyle` to get the full text styling
- [ ] add more assertions and cover more edge cases in the unit tests, especially for leap years

Wilker0725 commented on src/calendar/calendar-month.tsx:
```diff
@@ -0,0 +1,90 @@
+        setMonthDate(months);
+    };
+
+    if (monthDate.length === 1) return null;
```
Thanks for pointing out this one again.

Wilker0725 commented on src/calendar/calendar-month.tsx:
```diff
@@ -0,0 +1,90 @@
+import { CalendarMonthProps, VariantMonth } from "./types";
+import { CalendarHelper } from "src/util/calendar-helper";
+
+export type Month = typeof Months[number];
```
I will update it to typescript `type` way. since the `const Months = [..., ,...,]` is not more been use in other function.

I used const for the type as example below

`
export const Months = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
] as const;
`

Read it as 
`type Month = typeof Months[number];`

It similar as 
`
type Month = "January" |  "February", ...., ...., "December"
`

qroll commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,248 @@
+        performOnChangeHandler(value);
+    };
+
+    const handleMonthYearClick = (
```
following up on our earlier discussion, the `handleView` and `handleMonthYearClick` handlers can be simplified as a day/month/year selection results in a `calendarDate` change

the proposed change is to let the child components pass the date back to the parent. this also means we don't have to pass so many `data-*` attributes around
```
const handleDateChange = (date) => {
  // set selected date
  // update calendar date
}

<CalendarDay 
  onSelect={(date) => handleDateChange(date)} />
```

Wilker0725 replied:
wa. this is really clean code!

Wilker0725 commented on src/calendar/calendar-month.tsx:
```diff
@@ -0,0 +1,90 @@
+        const value = date.format("YYYY-MM-DD");
+        let variant: VariantMonth = "default";
+
+        if (!dayjs(value).isValid())
```
Arh! not any scenario will show the invalid for this calendar. This function is for the input with calendar

Wilker0725 replied:
I'll remove this `isValid` from the calendar first. Will add it back if needed from input calendar after merged.

Wilker0725 commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,39 @@
+export type VariantMonth =
+    | "default"
+    | "currentMonth"
+    | "nextMonth"
```
thanks for pointing out for this.

Wilker0725 commented on src/calendar/calendar-day.style.tsx:
```diff
@@ -0,0 +1,180 @@
+import { VariantDay } from "./types";
+
+interface DayNumerProps {
+    variant: VariantDay;
```
one question.

the classname also included?

qroll replied:
let me revisit the css styling. I missed out on this. I think it can be reworked so we don't have to use `className`, `data-cell`, `data-circle`. as Keith mentioned [here](https://github.com/LifeSG/react-design-system/pull/84#discussion_r1100966021), we should avoid passing down classnames and make use of props for styling.

qroll replied:
so the intention here is to style `Circle` based on whether this date is selected, ya?

instead of styling `Circle` indirectly from the parent which is `GrowDayCell`, you can pass an additional prop like `<Circle $selected={ /* logic to determine if date is selected */ } />`

now all the styles for `Circle` are in one place, which makes it easier to check and update in the future

Wilker0725 replied:
Yes and also not.

This calendar will be use in range selection later. The <LeftCell /> element handle Styling Left Side during responsive.

Left SideCell
Top, Bottom Dashed - start from LeftCell, so it wouldn't break in responsive. Because the LeftCell always will grow during responsive.

Wilker0725 replied:
removed data-attribute for selection from css.

replace with 
```
${props => {
  ${LeftCell} ${Circle} {}
}}
```

Wilker0725 commented:
Updated: 
1. Rename YearComponent to `renderYearHeader`
2. Remove handleView, and do it as toggleMonth and Year
3. Callback value from `CalendarDay`, `CalendarMonth` and `CalendarYear` 
4. Remove unused attribute from the Icon
5. Remove isValid. Look like it didn't been use in calendar. But it will be added in input calendar after merge if needed.
6. All data attribute with prefix `$` 
7. Updated all px to rounded Rem.
8. Create new ColorSet for Shadow. `${Color.Shadow.Accent}`
9. Rename some typeName inside typescript file. "nextMonth" & "nextDecade"

Pending:
1. regarding case "year" from your comment in this file `src/calendar/calendar.style.tsx`
2.classname in`<GrowDayCell />` required prefix with `$` also?

Wilker0725 commented:
Updated:

1. Remove ClassName and data-attribute
2. Unit Test

keithtxw commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,44 @@
+
+export type View = "Day" | "Month" | "Year";
+
+export type VariantDay = "default" | "otherMonth" | "today";
```
`DayVariant` and for the values to use kebab case

keithtxw commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,44 @@
+import { Dayjs } from "dayjs";
+
+export type View = "Day" | "Month" | "Year";
```
lowercase for consistency

keithtxw commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,44 @@
+
+export type VariantDay = "default" | "otherMonth" | "today";
+
+export type VariantMonth = "default" | "currentMonth" | "selectedMonth";
```
kebab case for consistency

keithtxw commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,44 @@
+
+export interface CalendarDayProps extends Pick<CalendarProps, "disabledDates"> {
+    selectedStartDate: string;
+    calendarDate: Dayjs;
```
i dont recommend you use a lib's type as the prop types. We cannot assume users to download the dependency. You should use either a string or `Date` object then do your own conversion to a `Dayjs` type instead.

keithtxw replied:
Apologies. This is not exposed, hence it is fine. Please ignore

keithtxw commented on src/util/calendar-helper.ts:
```diff
@@ -0,0 +1,66 @@
+        return years;
+    };
+}
+
```
Add a H1 header and indicate `HELPER FUNCTIONS`

keithtxw commented on src/util/calendar-helper.ts:
```diff
@@ -0,0 +1,66 @@
+import dayjs, { Dayjs } from "dayjs";
+
+export namespace CalendarHelper {
+    export const generateDay = (calendarDate: Dayjs): Dayjs[][] => {
```
`generateDays` plural for consistency

keithtxw commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,124 @@
+
+interface CalendarViewProps {
+    children: JSX.Element[];
+    $show: View;
```
rename to `$view`. if you use `$show` i would think it's a `boolean`

keithtxw commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,124 @@
+    border: 1px solid ${Color.Neutral[5]};
+    height: 25.125rem;
+    border-radius: 0.75rem;
+    padding-top: 2rem;
```
can we condense this to `padding: 2rem 0`

qroll commented on src/calendar/calendar-year.tsx:
```diff
@@ -0,0 +1,76 @@
+    onSelect,
+    onDecadeChange,
+}: CalendarYearProps) => {
+    const [yearDate, setYearDate] = useState<Dayjs[]>([]);
```
rename variable to just `years`

qroll commented on src/calendar/calendar-year.tsx:
```diff
@@ -0,0 +1,76 @@
+    }, [showView, calendarDate]);
+
+    const generateYearStatus = (date: Dayjs) => {
+        const grayYearIndex = [0, 11];
+
+        const isDecadeYear = grayYearIndex.includes(yearDate.indexOf(date));
```
can help to rename these variables to reflect their use
- `grayYearIndex` to `otherDecadeIndexes`
- `isDecadeYear` to `isOtherDecade`

qroll commented on src/calendar/calendar-year.tsx:
```diff
@@ -0,0 +1,76 @@
+
+        const isDecadeYear = grayYearIndex.includes(yearDate.indexOf(date));
+        const fullDate = date.format("YYYY-MM-DD");
+        const year = fullDate.split("-")[0];
```
this can be `date.year()` right?

Wilker0725 replied:
yes.

the FullDate (YYYY-MM-DD) needed also for below code.

isSame("YYYY-MM-DD") ----> format

```
   const variant: VariantYear = isOtherDecade
            ? "other-decade"
            : dayjs(selectedStartDate).isSame(fullDate, "year")
            ? "selected-year"
            : dayjs().isSame(fullDate, "year")
            ? "current-year"
            : "default";

```

Wilker0725 replied:
Updated to this

```
const fullDate = date.format("YYYY-MM-DD");
const year = date.year();
```

qroll commented on src/calendar/calendar-year.tsx:
```diff
@@ -0,0 +1,76 @@
+                return (
+                    <YearCell
+                        key={`${value}-${yearIndex}`}
+                        data-variant={variant}
```
`data-variant` is unused

qroll commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,124 @@
+    justify-content: center;
+    cursor: pointer;
+
+    p {
+        ${TextStyleHelper.getTextStyle("H4", "regular")}
+    }
```
Can extract this to a standalone `DropdownText` and replace `Text.BodySmall` in `calendar.tsx`

Wilker0725 commented:
Updated:

1. Typescript types.ts file only expose for user. 
2. Move types into file itself if not expose for user.
3. Renaming to kebab-case in type and lowercase
4. Update to plural for function naming

qroll commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,124 @@
+        }
+        return css`
+            ${styles}
+        `;
```
use map instead to improve readability. styled-components can handle array of styles. it's safe to assume there will always be 3 children
```
return ["day", "month", "year"].map((view, i) => `
  // css
`)
```

qroll commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,173 @@
+    // =============================================================================
+    const renderYearHeader = () => {
+        if (showView === "Year") {
+            const beginDecaded =
+                Math.floor(+calendarDate.format("YYYY") / 10) * 10;
+
+            const endDecaded = dayjs(`${beginDecaded}-01-01`)
+                .add(9, "year")
+                .format("YYYY");
```
please help to move this to the `CalendarHelper` and rename `Decaded` to `Decade`

btw I think endDecade can be calculated directly with `beginDecade + 9`

qroll commented on src/spec/color-spec/base-color-set.ts:
```diff
@@ -58,4 +58,7 @@ export const BaseColorSet: ColorSet = {
         },
     } as ValidationTypes,
+    Shadow: {
+        Accent: "#57A9FF80",
```
just a heads up that there will be a merge conflict for these color set files after #87 goes in

Wilker0725 commented on src/calendar/calendar-day.style.tsx:
```diff
@@ -0,0 +1,174 @@
+
+        if ($selected) {
+            return css`
+                ${LeftCell} ${Circle} {
```
Arh.. I tried to create LeftCircle but some error throw from styled component.

```
const LeftCircle = LeftCell > LeftCircle;
```

qroll replied:
oh, I meant creating new components. like `const LeftCircle = styled(Circle)`

Wilker0725 replied:
Arh. how we indicated that is Left?

Because the Circle will shared in 
```
<LeftCell>
  <LeftCircle />
</LeftCell />

<RightCell>
  <RightCircle />
</RightCell />
```

Wilker0725 replied:
is this your expected?

```
 ${LeftCell} ${LeftCircle} {
    background-color: ${Color.Accent.Light[5]};
    border: 1px solid ${Color.Primary};
}
```

qroll replied:
yep pretty much, but if you just want to style `LeftCircle`, then no need to specify `LeftCell`

Wilker0725 replied:
Let me know if i'm wrong.

I need to select which Circle i selected.

In Dom, there has two Circle, 
1. LeftCell Circle
2. RightCell Circle

If i just (below code), it selected both <Circle /> element.
```
 ${LeftCircle} {
    background-color: ${Color.Accent.Light[5]};
    border: 1px solid ${Color.Primary};
}
```

Wilker0725 replied:
Opps. Yes.

You are right and explanation. thank for your time!

The `LeftCircle` was already been created and use into the DOM to specific the element, so not really need the `LeftCell` to indicated the `Circle`. 

eg:

```
export const LeftCircle = styled(Circle)``;

 <LeftCell>
  <LeftCircle // ---------> specific use for LeftCircle even it inherit for Circle styled.
     $position="left"
     $variant={variant}
   />
</LeftCell>
```

Wilker0725 commented on src/calendar/calendar-month.style.tsx:
```diff
@@ -0,0 +1,60 @@
+
+    &:hover {
+        box-shadow: ${(props) =>
+            props.theme.name === "BookingSG"
```
oops, missed this to update.

Question:
i can update it now, right? any relation with #87  PR

qroll replied:
yes can update first

Wilker0725 commented:
Updated:

1. Add unit test for leap year in `generateMonths`. (not added leap year for generateDecadeOfYears, it will trigger the `generateMonths` once the value updated)
2. Add unit test for getStartEndDecade (new fn)
3. Update map loop in css instead for()
4.

Wilker0725 commented:
Pending:

1. found buggy in generateDecadeOfYears, will do a unit test for the leap year also

qroll commented on src/calendar/calendar-month.tsx:
```diff
@@ -47,7 +47,18 @@ export const CalendarMonth = ({
+            result = dayjs(`${yyyy}-${selectedMonth}-${dd}`);
+        }
+
+        onSelect(result);
```
is this bug because generateMonths and generateDecadeOfYears use calendarDate instead of selectedStartDate?

Wilker0725 replied:
Yes, kind of.

We have two scenario:
1. User didn't select day in calendar but selected Month or Year first -> we need to selected default date and update accordingly if user update for month.
2. User selected day and update month calendar, so need to know some value been selected.

Wilker0725 replied:
I'll update the year calendar fn next day. I spotted one bug. 

Selected 2024 Feb 29 -> update to -> 2023 Feb -> it is not selected last day of Feb.

qroll replied:
ok. I think instead of fixing it in handleMonthClick, should fix the date generation instead

Wilker0725 replied:
Thanks giving some clue. let me think. :))

Wilker0725 replied:
`generateMonths` which a function to generate calendar months.

All month value will be same as selected value in the calendar for prev/next year if we update the selected value inside this value.

Wilker0725 commented:
Updated:

1. Remove last `handleMonthClick` updated. Update the logics inside `generateFn`  
2. Add unit test for leap year in `generateDecadeOfYears`

Wilker0725 commented:
Updated:
1. rebase with master
2. update last day checking inside month/year generateFn
3. update unit test

Wilker0725 commented:
Updated:

1. Remove begin `0` for the calendar. eg: 04 -> should show -> 4 in the calendar
2. Clean up generateFn 
3. Unit test for selected lastday if the month/year didn't exist the day

qroll commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,167 @@
+                    <DropdownText>
+                        {dayjs(calendarDate).format("MMM")}
+                    </DropdownText>
+                    <ChevronDownIcon />
```
looks larger in Figma, try styling it with height and width 1.125rem

qroll commented on src/calendar/calendar-day.style.tsx:
```diff
@@ -0,0 +1,176 @@
+
+const Circle = styled.div<CircleProps>`
+    ${TextStyleHelper.getTextStyle("H5", "regular")}
+    color: ${Color.Neutral[1]};
```
is color needed? seems to be handled in `DayNumber`

Wilker0725 replied:
Thank you so much for your time to understand my code what exactly does.

qroll commented on src/calendar/calendar-month.style.tsx:
```diff
@@ -0,0 +1,57 @@
+    display: grid;
+    grid-template-columns: repeat(2, 13.5rem);
+    grid-template-rows: repeat(6, 2.5rem);
+    gap: 0.5rem 1rem;
```
space between columns seems wider in figma
```suggestion
    gap: 0.5rem 2.5rem;
```

qroll commented on src/util/calendar-helper.ts:
```diff
@@ -1,9 +1,6 @@
 export namespace CalendarHelper {
-    // =============================================================================
-    // HELPER FUNCTIONS
-    // =============================================================================
```
might have miscommunicated - this header was meant to be moved down, above the unexported functions

qroll reviewed (commented):
found some small styling discrepancies with Figma

Wilker0725 commented on src/calendar/calendar-day.style.tsx:
```diff
@@ -0,0 +1,176 @@
+export const LeftCircle = styled(Circle)``;
+export const RightCircle = styled(Circle)``;
+
+export const DayNumber = styled.div<DayNumberProps>`
```
opps. thanks for your time. 

I guess i removed it accidentally. let me check all the styles.

Wilker0725 commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,119 @@
+    top: 50%;
+    transform: translateY(-50%);
+    color: ${Color.Neutral[3]};
+    cursor: pointer;
```
I created a `type` for this calendar for ready styling in future input 

`$type: 'calendar' | 'input'`

Wilker0725 commented:
Updated:

1. Create new props `$type` for calendar, this will be use in input calendar later.
2. Update `today` selection color and font weight.
3. Update `selected` color and font weight.
4. Update `grid gap` in MonthComponent and YearComponent

keithtxw commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,159 @@
+    cursor: pointer;
+`;
+
+export const DropdownText = styled(Text.BodySmall)`
```
Just use `<Text.H4 weight="regular">`

keithtxw commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,159 @@
+export const YearView = styled(ViewSectionBase)``;
+
+const ArrowLeftRightBase = styled(ChevronLeftIcon)`
+    position: absolute;
```
I would recommend extracting common styles to a function, then use the correct icon rather than doing transformations (e.g. `ArrowRight`)

keithtxw commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,172 @@
+    // CONST, STATE, REF
+    // =============================================================================
+    const [calendarDate, setCalendarDate] = useState<Dayjs>(dayjs());
+    const [showView, setShowView] = useState<View>("day");
```
can rename to `currentView`, `showView` suits a boolean value more

keithtxw commented on src/calendar/calendar-year.tsx:
```diff
@@ -0,0 +1,79 @@
+import { CalendarHelper } from "../util/calendar-helper";
+import { YearMonthBase } from "./calendar-month";
+
+export type VariantYear =
```
Can i suggest renaming to `YearVariant`

keithtxw commented on src/calendar/calendar-day.tsx:
```diff
@@ -0,0 +1,126 @@
+} from "./calendar-day.style";
+import { CalendarProps } from "./types";
+
+export interface DayCellStyleProps {
```
this should be in the style file

keithtxw commented on src/calendar/calendar-day.tsx:
```diff
@@ -0,0 +1,126 @@
+    $disabled?: boolean;
+}
+
+export type VariantDay = "default" | "other-month" | "today";
```
Can i request it to be renamed to `DayVariant`

keithtxw commented on src/calendar/calendar-year.tsx:
```diff
@@ -0,0 +1,79 @@
+    | "other-decade"
+    | "selected-year";
+
+interface CalendarYearProps extends YearMonthBase {}
```
while it seems cleaner, i think it is better to separate the interface for clarity and unnecessary export

keithtxw commented on src/calendar/calendar-year.style.tsx:
```diff
@@ -0,0 +1,56 @@
+import { TextStyleHelper } from "../text/helper";
+import { VariantYear } from "./calendar-year";
+
+interface YearCellProps {
```
Rename to `StyleProps`

keithtxw commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,159 @@
+import { MonthPickerContainer } from "./calendar-month.style";
+import { YearPickerContainer } from "./calendar-year.style";
+
+interface ContainerProps extends Pick<CalendarProps, "$type"> {}
```
rename `ContainerStyleProps`

keithtxw commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,159 @@
+}
+
+interface CalendarViewProps {
+    children: JSX.Element[];
```
why do you need this prop?

Wilker0725 replied:
I will removed this `children` prop, the typescript not complaint anymore. :'))

keithtxw commented on stories/calendar/calendar.stories.mdx:
```diff
@@ -0,0 +1,32 @@
+
+<Secondary>Overview</Secondary>
+
+Displays the calendar with today date in default.
```
Displays a calendar component.

keithtxw commented on stories/calendar/props-table.tsx:
```diff
@@ -0,0 +1,47 @@
+                propTypes: ["string"],
+            },
+            {
+                name: "onChange",
```
I tend to put the event handlers at the bottom

keithtxw commented on stories/calendar/props-table.tsx:
```diff
@@ -0,0 +1,47 @@
+                        <code>YYYY-MM-DD</code>
+                    </>
+                ),
+                propTypes: ["string"],
```
you're missing the default which is today?

Wilker0725 replied:
??

the `value` props for the value from the server. if there is not exist, not selected the calendar after component mounted.

keithtxw commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,6 @@
+export interface CalendarProps {
+    disabledDates?: string[] | undefined;
```
add jsdoc comment to show an example format

qroll replied:
the comment should be directly above the prop it is describing. you can reference `src/error-display/types.ts`

keithtxw commented on src/calendar/types.ts:
```diff
@@ -0,0 +1,6 @@
+export interface CalendarProps {
+    disabledDates?: string[] | undefined;
+    onChange?: ((value: string) => void) | undefined;
+    value?: string | undefined; // YYYY-MM-DD
```
add jsdoc to show example format

keithtxw replied:
we'll be also adding for the other relevant components

Wilker0725 commented:
Updated:

1. Add respective headers in components
2. Rename style element and types. 
3. Separately the interface in component file, no more export. (month and year)
4. Add jsdoc comment for type in types.tx

keithtxw commented on stories/calendar/props-table.tsx:
```diff
@@ -0,0 +1,58 @@
+                        <code>YYYY-MM-DD</code>
+                    </>
+                ),
+                propTypes: ["YYYY-MM-DD"],
```
This should be a `string[]`

Wilker0725 replied:
can use string in props-table.tsx file?

keithtxw replied:
I mean the type is `"string[]"`

keithtxw replied:
I will make the amendments

keithtxw commented on stories/calendar/props-table.tsx:
```diff
@@ -0,0 +1,58 @@
+                name: "disabledDates",
+                description: (
+                    <>
+                        Date value will be disabled, unselectable.&nbsp;
```
The dates to be disabled. E.g. <code>[YYYY-MM-DD, YYYY-MM-DD]</code>

keithtxw commented on tests/util/__snapshots__/calendar-helper.spec.ts.snap:
```diff
@@ -0,0 +1,94 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Don't need snapshots for functions test

keithtxw replied:
Understand. We seldom do this. But if your intention is to "freeze" the expected outputs, then I guess it is fine.

keithtxw commented on src/calendar/calendar.tsx:
```diff
@@ -0,0 +1,179 @@
+                </DropdownYear>
+            </HeaderDropdown>
+            <ContentBody>
+                <ArrowLeft onClick={handleLeftArrowClick} />
```
I prefer wrapping this in a button for semantic and accessibility purpose

keithtxw commented on src/calendar/calendar.style.tsx:
```diff
@@ -0,0 +1,160 @@
+export const YearView = styled(ViewSectionBase)``;
+
+export const ArrowLeft = styled(ChevronLeftIcon)`
+    color: ${Color.Neutral[3]};
```
extract out into a common style to be used by both icons

Wilker0725 commented:
Updated:

1. `IconButton` to wrap the `Icon`
2. Create a css style shared with `ArrowLeft` & `ArrowRight`
3. Editing the table props desc

Note:
1. The `string[]`, i didn't update it yet. please help. :))
